File: iosApp/iosApp/appwrite/AppwriteDatabaseApple.swift
//
//  AppwriteDatabaseApple.swift
//  iosApp
//
//  Created by Stanislav Aleshin on 26.06.2025.
//  Copyright © 2025 orgName. All rights reserved.
//
import Foundation
import shared
import Appwrite
import AppwriteModels
import JSONCodable


public class AppwriteDatabaseApple : RemoteAppwriteDatabaseApple {
    
    private var database: Databases
    
    init(database: Databases) {
        self.database = database
    }
    
    
    public override func __createDocument(
        databaseId: String,
        collectionId: String,
        documentId: String,
        data: Any,
        permissions: [String]?
    ) async throws -> RemoteDocument<NSDictionary> {
        let result: AppwriteModels.Document<[String: AnyCodable]> = try await database.createDocument(
           databaseId: databaseId,
           collectionId: collectionId,
           documentId: documentId,
           data: data,
           permissions: permissions
        )
        return RemoteDocument.init(
            id: result.id as String,
            collectionId: result.collectionId,
            databaseId: result.databaseId,
            createdAt: result.createdAt,
            updatedAt: result.updatedAt,
            permissions: result.permissions,
            data: convertToKotlinCompatibleDictionary(result.data)
        )
       }

       public override func __deleteDocument(
            databaseId: String,
            collectionId: String,
            documentId: String
       ) async throws -> Any {
           return try await database.deleteDocument(
               databaseId: databaseId,
               collectionId: collectionId,
               documentId: documentId
           )
       }


       public override func __getDocument(
            databaseId: String,
            collectionId: String,
            documentId: String,
            queries: [String]?
       ) async throws -> RemoteDocument<NSDictionary> {
           let result: AppwriteModels.Document<[String: AnyCodable]> = try await database.getDocument(
               databaseId: databaseId,
               collectionId: collectionId,
               documentId: documentId,
               queries: queries
           )
           return RemoteDocument(
                id: result.id as String,
                collectionId: result.collectionId,
                databaseId: result.databaseId,
                createdAt: result.createdAt,
                updatedAt: result.updatedAt,
                permissions: result.permissions,
                data: convertToKotlinCompatibleDictionary(result.data)
           )
       }

       public override func __listDocuments(
            databaseId: String,
            collectionId: String,
            queries: [String]?
       ) async throws -> RemoteDocumentList<NSDictionary> {
           let result: AppwriteModels.DocumentList<[String: AnyCodable]> = try await database.listDocuments(
               databaseId: databaseId,
               collectionId: collectionId,
               queries: queries
           )
           let mapped = result.documents.map {
               RemoteDocument(
                    id: $0.id as String,
                    collectionId: $0.collectionId,
                    databaseId: $0.databaseId,
                    createdAt: $0.createdAt,
                    updatedAt: $0.updatedAt,
                    permissions: $0.permissions,
                    data: $0.data as NSDictionary
               )
           }
           return RemoteDocumentList(total: Int64(result.total), documents: mapped)
       }

       public override func __updateDocument(
            databaseId: String,
            collectionId: String,
            documentId: String,
            data: Any?,
            permissions: [String]?
       ) async throws -> RemoteDocument<NSDictionary> {
           let result: AppwriteModels.Document<[String: AnyCodable]> = try await database.updateDocument(
               databaseId: databaseId,
               collectionId: collectionId,
               documentId: documentId,
               data: data,
               permissions: permissions
           )
           return RemoteDocument(
                id: result.id as String,
                collectionId: result.collectionId,
                databaseId: result.databaseId,
                createdAt: result.createdAt,
                updatedAt: result.updatedAt,
                permissions: result.permissions,
                data: convertToKotlinCompatibleDictionary(result.data)
           )
       }

       public override func __upsertDocument(
            databaseId: String,
            collectionId: String,
            documentId: String,
            data: Any,
            permissions: [String]?
       ) async throws -> RemoteDocument<NSDictionary> {
           let result: AppwriteModels.Document<[String: AnyCodable]> = try await database.upsertDocument(
               databaseId: databaseId,
               collectionId: collectionId,
               documentId: documentId,
               data: data,
               permissions: permissions
           )
           return RemoteDocument(
                id: result.id as String,
                collectionId: result.collectionId,
                databaseId: result.databaseId,
                createdAt: result.createdAt,
                updatedAt: result.updatedAt,
                permissions: result.permissions,
                data: convertToKotlinCompatibleDictionary(result.data)
           )
       }
}

File: shared/core/remote/src/iosMain/kotlin/ru/aleshin/studyassistant/core/remote/appwrite/AppwriteDatabaseApple.kt
/*
 * Copyright 2024 Stanislav Aleshin
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ru.aleshin.studyassistant.core.remote.appwrite

import ru.aleshin.studyassistant.core.remote.appwrite.databases.AppwriteDatabase
import ru.aleshin.studyassistant.core.remote.appwrite.databases.Document
import ru.aleshin.studyassistant.core.remote.appwrite.databases.DocumentList
import ru.aleshin.studyassistant.core.remote.appwrite.databases.mapData
import kotlin.reflect.KClass

/**
 * @author Stanislav Aleshin on 28.06.2025.
 */
abstract class AppwriteDatabaseApple : AppwriteDatabase {

    override suspend fun <T : Any> listDocuments(
        databaseId: String,
        collectionId: String,
        queries: List<String>?,
        nestedType: KClass<T>
    ): DocumentList<T> {
        val documentList = listDocuments(
            databaseId = databaseId,
            collectionId = collectionId,
            queries = queries,
        )

        return DocumentList(
            total = documentList.total,
            documents = documentList.documents.map { it.mapData(nestedType) },
        )
    }

    override suspend fun <T : Any> createDocument(
        databaseId: String,
        collectionId: String,
        documentId: String,
        data: Any,
        permissions: List<String>?,
        nestedType: KClass<T>
    ): Document<T> {
        val document = createDocument(databaseId, collectionId, documentId, data, permissions)
        return document.mapData(nestedType)
    }

    override suspend fun <T : Any> getDocument(
        databaseId: String,
        collectionId: String,
        documentId: String,
        queries: List<String>?,
        nestedType: KClass<T>
    ): Document<T> {
        val document = getDocument(databaseId, collectionId, documentId, queries)
        return document.mapData(nestedType)
    }

    override suspend fun <T : Any> upsertDocument(
        databaseId: String,
        collectionId: String,
        documentId: String,
        data: Any,
        permissions: List<String>?,
        nestedType: KClass<T>
    ): Document<T> {
        val document = upsertDocument(databaseId, collectionId, documentId, data, permissions)
        return document.mapData(nestedType)
    }

    override suspend fun <T : Any> updateDocument(
        databaseId: String,
        collectionId: String,
        documentId: String,
        data: Any?,
        permissions: List<String>?,
        nestedType: KClass<T>
    ): Document<T> {
        val document = updateDocument(databaseId, collectionId, documentId, data, permissions)
        return document.mapData(nestedType)
    }
}